homeassistant/components/media_player/dlna_dmr.py

```diff
@@ -0,0 +1,717 @@
+            return func(self, action, *args, **kwargs)
+
+        return wrapper
+
+    return call_wrapper
+
+
+def requires_state_variable(service_type,
+                            state_variable_name,
+                            value_not_connected=None):
+    """
+    Ensure service/state_variable is available.
+
+    If not available, then raise NotImplemented.
+    """
+    def call_wrapper(func):
+        """Call wrapper for decorator."""
+        @functools.wraps(func)
+        def wrapper(self, *args, **kwargs):
+            """
+            Require device is connected and has service/state_variable.
+
+            If device is not connected, value_not_connected is returned.
+            """
+            # pylint: disable=protected-access
+            if not self._is_connected:
+                return value_not_connected
+
+            service = self._service(service_type)
+            if not service:
+                _LOGGER.error('requires_state_variable(): '
+                              '%s.%s: no service: %s',
+                              self,
+                              func.__name__, service_type)
+                raise NotImplementedError()
+
+            state_var = service.state_variable(state_variable_name)
+            if not state_var:
+                _LOGGER.error('requires_state_variable(): '
+                              '%s.%s: no state_variable: %s.%s',
+                              self,
+                              func.__name__,
+                              service_type,
+                              state_variable_name)
+                raise NotImplementedError()
+            return func(self, state_var, *args, **kwargs)
+        return wrapper
+    return call_wrapper
+
+
+def start_notify_view(hass):
+    """Register notify view."""
+    hass_data = hass.data[__name__]
+    name = 'notify_view'
+    if name in hass_data:
+        return hass_data[name]
+
+    view = UpnpNotifyView(hass)
+    hass_data[name] = view
+    hass.http.register_view(view)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up DLNA DMR platform."""
+    if discovery_info and \
+       'upnp_device_type' in discovery_info and \
+       discovery_info['upnp_device_type'] != UPNP_DEVICE_MEDIA_RENDERER:
+        _LOGGER.debug('Device is not a MediaRenderer: %s',
+                      discovery_info.get('ssdp_description'))
+        return
+
+    if config.get(CONF_URL) is not None:
+        url = config.get(CONF_URL)
+        name = config.get(CONF_NAME)
+        udn = config.get(CONF_UDN)
+    elif discovery_info is not None:
+        url = discovery_info['ssdp_description']
+        name = discovery_info['name']
+        udn = discovery_info['udn']
+
+    # set up our Views, if not already done so
+    if __name__ not in hass.data:
+        hass.data[__name__] = {}
+
+    hass.async_run_job(start_notify_view, hass)
+    notify_view = hass.data[__name__]['notify_view']
```

Comment:
I'm getting an exception at this point, most likely that async job should be a synchronous one, or this variable should first be accessed after it has signaled to be ready:
```
Jun 30 17:49:02 nuc hass[18016]:   File "/home/hass/home-assistant/homeassistant/components/media_player/dlna_dmr.py", line 204, in setup_platform
Jun 30 17:49:02 nuc hass[18016]:     notify_view = hass.data[__name__]['notify_view']
Jun 30 17:49:02 nuc hass[18016]: KeyError: 'notify_view'
```